### ISSUE TITLE
Epic bug: Please support my favorite package manager!

### ISSUE THREAD

**Issue by bitspittle** [varabyte/kobweb#117]
In order to avoid everyone creating one off bugs for the same package manager, this bug will act as a home to collect all such requests received so far. (Of course, once someone starts working on adding support for one of them, that could be a good time to create an individual bug).

Here is the current list of package managers we are considering (but not actively working on):

* apt
* chocolatey
* gofish
* macports
* snapcraft
* spec

---

**NOTE**: **Some of these package managers may never get supported**, as I'm not sure how many people actually want to use them, and supporting more releases results in more work on our end. So if you see your package manager on this list, it still may be worth leaving a comment below like "Yes, I want that one please! I use it particularly because..."

**Comment by russellbanks** [comment]
It would be nice to have this on Winget! Currently however, [Winget doesn't accept zip files](https://github.com/microsoft/winget-cli/issues/140) but rather installers so we'd have to either wait until they do support zip files (it's an open issue but there is no indication of this being implemented) or create an executable (exe or msi).

**Comment by bitspittle** [comment]
Thank you @russellbanks for your comment. I updated the initial list to include winget.

Hmmm, it might be possible to create an exe... but at that point I think I would be bundling a jre which might balloon the final size a bit, which may be unfortunate since to use Kobweb the user pretty has to have Java installed on their machine already. 

Hopefully, though, winget decided to support .zip files. It is so convenient and would reduce so much friction!

**Comment by russellbanks** [comment]
I'm not the most well-versed in packaging Java applications (in terms of bundling the JRE, etc) but is it possible to create an executable without bundling the JRE? If this is not feasible, we could just bite the bullet and bundle the JRE with it, specifying that the recommended option is still the zip file?

Even if this is a temporary solution, I still think I would be advantageous to look into ways we can get an executable working in order to get this on Winget as it is the official package manager for Windows.

P.S. Just a correction to the updated initial list - the link for Winget is [https://github.com/microsoft/winget-cli](https://github.com/microsoft/winget-cli).

**Comment by bitspittle** [comment]
@russellbanks Thanks for letting me know about the wrong link. I fixed the original comment.

Even if we could build an .exe, just keep in mind I'd want a flow where this could all be done automatically every time I released a new version. With existing package managers, this is basically a one click operation for all of them so far. So we'd need to figure out how to automatically 1) unzip the file, 2) bundle it up into an exe, 3) and publish this to winget.

I personally won't have time to look into this for quite a while, but if someone knows how to do it, and/or has an example of this happening already, please add a comment -- that could help a lot!

P.S. Russell, you may want to check out the https://github.com/jreleaser/jreleaser project and file an issue with them to support winget. I'm relying on them pretty heavily for my other releases, and if winget is the official package manager for windows, they may be interested in getting some signal that people would want it supported.

**Comment by russellbanks** [comment]
Not much more I can comment on this other than I hope that either Winget supports zip files in the future or JReleaser adds support for Winget. There is an [open issue](https://github.com/jreleaser/jreleaser/issues/185) on JReleaser already to ask for support for winget as a packager, but it doesn't seem to have any activity.

**Comment by vedantmgoyal2009** [comment]
JReleaser now has an MSI installer which can be published to WinGet. I think we can now proceed.

**Comment by bitspittle** [comment]
Great, thanks for the update! With JReleaser leading the way, I will
definitely take a look.

**Comment by russellbanks** [comment]
As of [1.4](https://github.com/microsoft/winget-cli/releases/tag/v1.4.2011-preview), winget will have support for installing zip files. However, this is just a preview at the moment, with the stable version being 1.2, and the release candidate being 1.3. But, this means that Kobweb can be put on WinGet; it's just a matter of time.

**Comment by aalmiray** [comment]
FWIW you can use `jpackage` (on its own or via JReleaser) to create Windows installers (exe, msi) that may be used in combination with Winget.

**Comment by bitspittle** [comment]
(Sorry for the recent radio silence, folks. I just moved, which wiped out my schedule for a couple of weeks).

As I catch up on stuff, I do intend to look into Winget, although I have a few other priorities to get through first. I'm hoping to get to this by end of September ('22) at the very latest, so feel free to ping this bug if it's later with no progress.

**Comment by bitspittle** [comment]
*Finally*, I'm back to this bug and looking into adding winget support. Thanks again for everyone's patience.

@aalmiray I'm trying to follow the Gradle syntax for jpackage from here: https://jreleaser.org/guide/latest/reference/assemble/jpackage.html

but "jpackage" contents aren't autocompleting for me:
![jreleaser-jpackage](https://user-images.githubusercontent.com/43705986/224174064-0c3d3e99-7002-4d40-80d7-bb54f28f460b.png)

(I'm using latest, Jreleaser v1.5.0)

is there any demo code for using jreleaser + jpackage I can look at? 

Also, according to @russellbanks , winget should support .zip files now ([1.4](https://github.com/microsoft/winget-cli/releases/tag/v1.4.10173) was released in Jan). Are there plans to add support for that into JReleaser, or will `jpackage`ing be the way to go for a while?

**Comment by russellbanks** [comment]
@bitspittle

https://github.com/jreleaser/jreleaser/issues/185

https://jreleaser.org/guide/latest/reference/packagers/winget.html#_distribution_support

**Comment by bitspittle** [comment]
Right, that's what I'm looking at. `winget` doesn't support .zip file targets according to that distribution support list, so I assume at this point I still need to try running `jpackage` to create an .exe for now.

**Comment by aalmiray** [comment]
Zip support in Winget was added in `1.4` but the team has decided to wait until a critical mass of users has migrated to `1.4`. See https://github.com/jreleaser/jreleaser/issues/185#issuecomment-1406908541

Once that happens we'll support releasing zips with Winget.

re: auto-completion. `app` is a free name which you can choose to be anythign you want. It identifies a named element inside a NamedContainer, thus you must use regular KotlinDSL such as

```kotlin
jreleaser {
  assemble {
    jpackage {
      create("kobweb-installer") {
        ...
      }
    }
  }
}
```

**Comment by bitspittle** [comment]
Ah perfect, it was a Kotlin script thing. Thanks for the quick reply!

**Comment by russellbanks** [comment]
> Zip support in Winget was added in `1.4` but the team has decided to wait until a critical mass of users has migrated to `1.4`. See [jreleaser/jreleaser#185 (comment)](https://github.com/jreleaser/jreleaser/issues/185#issuecomment-1406908541)

@aalmiray zip's were turned in on the pipelines a month or so ago:
https://github.com/microsoft/winget-pkgs/issues/87300

**Comment by aalmiray** [comment]
@russellbanks gotcha. Will have a look at zips on Winget in the next couple of weeks.

**Comment by aalmiray** [comment]
Filed a new ticket to keep track of zip support on Winget https://github.com/jreleaser/jreleaser/issues/1253

**Comment by bitspittle** [comment]
I think I'll hold off on winget until zip support is ready.

It looks like supporting building an MSI/EXE installer will be rather tricky, involving a bunch of structure just for winget that no other package manager has needed. For example, to run `jlink / jpackage`, I think I would need to set up a special CI workflow using Java 16, since everything else I've done works fine with JDK11 (or older). 

And given the extra complexity around this, I have other priorities I should attend to first.

**Comment by aalmiray** [comment]
Understood. If it helps at a later stage, have a look a JReleaser's [release workflow](https://github.com/jreleaser/jreleaser/blob/main/.github/workflows/release.yml) and [configuration](https://github.com/jreleaser/jreleaser/blob/main/jreleaser.yml) which take care of multiple targets during a release.

**Comment by bitspittle** [comment]
Thanks @russellbanks , I appreciate the offer!

However, I'd want to make sure that everything is handled in the build script so whenever I update the Kobweb binary, I can do everything with a single `jreleaserPublish`.

Probably best to figure everything out the correct way all at once (or otherwise I might release future versions of Kobweb and winget would fall behind)

**Comment by russellbanks** [comment]
Anything you do do to automate winget should take the previous manifest and update the values with that of your new artifacts, so a pre-existing manifest shouldn't stop you doing that in the future.

**Comment by bitspittle** [comment]
@aalmiray Yeah I was using one of your simple workflows as a guide, but seeing that I'd need  to set up a bunch of runtime JDKs just for winget seems like a lot of work.

Also, I can just jlink / jpackage a windows binary right? If I don't care about the other platforms?

And if I am building an installer for windows, do I need to make sure to run jreleaser on a Windows machine?

**Comment by bitspittle** [comment]
@russellbanks where does the winget manifest live? Currently, I have some local repositories (e.g. for homebrew) which jreleaser overwrites for me. Is winget different? (I was skimming the docs but haven't looked too closely yet)

**Comment by aalmiray** [comment]
> Also, I can just jlink / jpackage a windows binary right? If I don't care about the other platforms?
Correct.

> And if I am building an installer for windows, do I need to make sure to run jreleaser on a Windows machine?
Correct, again.

**Comment by russellbanks** [comment]
> @russellbanks where does the winget manifest live? Currently, I have some local repositories (e.g. for homebrew) which jreleaser overwrites for me. Is winget different? (I was skimming the docs but haven't looked too closely yet)

[microsoft/winget-pkgs](https://github.com/microsoft/winget-pkgs)

Pull requests are just made to winget-pkgs, usually using and created with a tool like [Komac](https://github.com/russellbanks/Komac), [YamlCreate](https://github.com/microsoft/winget-pkgs/blob/master/Tools/YamlCreate.ps1) or [WinGetCreate](https://github.com/microsoft/winget-create).

**Comment by bitspittle** [comment]
@russellbanks Ah thanks, really helpful info! I don't think I'd mind you submitting an initial manifest for me, then, if it doesn't block me from updating or overwriting it later.

I was trying to search randomly through the manifests.... do you know of a manifest out there which uses the new zip type?

FYI latest Kobweb zip is here: https://github.com/varabyte/kobweb/releases/tag/cli-v0.9.11

However, it's not a standalone zip file. The user will need to have Java 11 (or newer) installed on their machine. 

Does winget support the idea of specifying "Java 11" as a dependency, at which point it handles installing things automatically? Or would it be up to me to bundle Java 11 into the zip file?

@aalmiray Does this mean when I run `jreleaser` targeting winget, it won't immediately update anything but will submit a PR on my behalf?

**Comment by russellbanks** [comment]
Ah, winget doesn't support dependencies currently. It is planned as a "soon" feature.
See [Dependencies.md on winget-cli](https://github.com/microsoft/winget-cli/blob/73f05cdf9b72eed991ccc5609f459f385cb9ec2f/doc/specs/%23163%20-%20Dependencies.md).

For the time being, you could make a second zip that has a stripped JRE using JLink bundled with it? That's entirely up to you though.

**Comment by aalmiray** [comment]
> @aalmiray Does this mean when I run jreleaser targeting winget, it won't immediately update anything but will submit a PR on my behalf?

Not by default, no. The original idea was to push to an intermediate repository (`winget-kobweb` for example) and trigger a GH workflow that pull create a branch and push. The PR would still have to be done manually _unless_ the workflow were to use the `gh` command. This is how https://github.com/jreleaser/jreleaser-macports is handled.

However, I just added an option to specify a `branchPush` property on taps/repositories (https://github.com/jreleaser/jreleaser/issues/1129) which would give you the ability to directly create a branch on a target repository such as `winget-pkgs`, bypassing `winget-kobweb`. The actual PR would still have to be created manually at the moment.

**Comment by bitspittle** [comment]
Hmmm, I'm thinking I should probably should look into jlink at this point. 

Give me a week or so, I have some other things I need to focus on first, but I'll ramp up on jlink / jpackage over the next few days.

Maybe I'll even give in and finally move publishing my artifacts via Jreleaser from a CI instead of my personal desktop, like an actual professional or something :stuck_out_tongue: Once that's migrated, then it wouldn't be too hard to add a workflow that runs on Windows for Winget.

(I just spent a week getting CI artifact support working for two of my other projects. It's a pain, to be honest, and I'm not looking forward to the game of CI trial and error yet again at the moment)

@russellbanks No `createDistributable` or anything like it in my tasks, sadly. Note that it seems pretty unlikely for Kobweb users to *not* have Java installed on their machine, since they're doing Kotlin development anyway. Having winget handle the dependencies for me would be the best though, I think.

**Comment by aalmiray** [comment]
> (I just spent a week getting CI artifact support working for two of my other projects. It's a pain, to be honest, and I'm not looking forward to the game of CI trial and error yet again at the moment)

That's why JReleaser has a `dry-run` mode and offers plenty of options to run as much as possible on a local environment. Alas, when platform specific binaries or tools are required you must test on the given platform.

**Comment by Skaldebane** [comment]
Hi there! It would be nice to support `.rpm` packages for Fedora and other RHEL-based Linux distros. Thanks!

**Comment by Skaldebane** [comment]
Flatpak is also a pretty solid one. It's ships with almost every Linux distro by default, and Flathub allows publishing CLI apps (although they won't be shown in the apps store understandably).

**Comment by bitspittle** [comment]
@Skaldebane thanks! I updated the original comment.

Do you have any experience with either of those package managers? At one point, I checked out snap and it was a wasted day.

**Comment by Skaldebane** [comment]
I've previously created an .rpm for a specific purpose, but nothing really proper.
As for Flatpak, I do have some experience with it (for making a GTK app), but not in the Java ecosystem.

However, I'll try to package Kobweb into one of them when I find some time!

**Comment by aalmiray** [comment]
FWIW JReleaser supports both packagers.

The Native Installer assembler relies on Jpackage to create an .rpm that bundles a Java Runtime along with the app. 

The Spec packager OTOH builds an .rpm as expected by Fedora COPR’s build infrastructure. 

Lastly, Flatpak may be used to package CLI and/or desktop apps

**Comment by bitspittle** [comment]
(Ah sorry @aalmiray meant to reply last week!)

Just wanted to say thanks for the last comment. If we can do everything with JReleaser, that's definitely the way to go.

@Skaldebane if you were going to try to experiment with this yourself, the jreleaser config / kobweb binary is actually in a different project. [You can find it here](https://github.com/varabyte/kobweb-cli/blob/2be5ac3b2eeb90ca2ca9c735f5b8cc75ce53dd28/kobweb/build.gradle.kts#L73).

**Comment by bitspittle** [comment]
Thanks for the reminder @Ayfri . I added it to my near-term todo list to revisit, to see if it will work seamlessly now with latest JReleaser.

I probably have a week or so more of Kobweb work to get through, and then I'll see if I can get winget support working after that. I'll report back here whether it goes well or if there are still obstacles.

**Comment by Ayfri** [comment]
Haha I see, no problem, unforeseen events happen to anyone !
